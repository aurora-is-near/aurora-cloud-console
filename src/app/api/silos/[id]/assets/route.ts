import { snakeCase } from "change-case"
import { abort } from "@/utils/abort"
import { toError } from "@/utils/errors"
import { createApiEndpoint } from "@/utils/api"
import { createStorageClient } from "@/supabase/create-storage-client"
import { updateSilo } from "@/actions/silos/update-silo"
import { SiloAsset } from "@/types/assets"

export const POST = createApiEndpoint("uploadSiloAsset", async (_req, ctx) => {
  const storage = createStorageClient()
  const file = ctx.body.get("file") as File
  const type = ctx.body.get("type") as SiloAsset

  if (!file) {
    abort(400, "No file provided")
  }

  const filename = snakeCase(file.name)
  const path = `/${ctx.params.id}/${type}sss/${filename}`

  const uploadRes = await storage.from("silo_assets").upload(path, file, {
    contentType: file.type,
    upsert: true,
  })

  if (uploadRes.error) {
    abort(500, toError(uploadRes.error).message)
  }

  const {
    data: { publicUrl: url },
  } = storage.from("silo_assets").getPublicUrl(path)

  await updateSilo(Number(ctx.params.id), {
    [type]: url,
  })

  return { url }
})
