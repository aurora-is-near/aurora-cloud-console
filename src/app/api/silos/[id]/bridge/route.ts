import { createApiEndpoint } from "@/utils/api"
import { abort } from "../../../../../utils/abort"
import { getSilo } from "@/actions/silos/get-silo"
import { getSiloBridge } from "@/actions/silo-bridge/get-silo-bridge"
import { createSiloBridge } from "@/actions/silo-bridge/create-silo-bridge"
import { Bridge, BridgeNetworkType } from "@/types/types"
import { BridgeSchema } from "@/types/api-schemas"
import { updateSiloBridge } from "@/actions/silo-bridge/update-silo-bridge"
import { isValidNetwork } from "@/utils/bridge"

const getValidNetworks = (networks: string[]): BridgeNetworkType[] => {
  const validNetworks = networks.filter(isValidNetwork)
  const invalidNetworks = networks.filter((network) => !isValidNetwork(network))

  if (invalidNetworks.length) {
    abort(400, `Invalid network(s): ${invalidNetworks.join(", ")}`)
  }

  return validNetworks
}

const getBridgeSchema = (bridge?: Bridge): BridgeSchema => {
  if (!bridge) {
    return {
      enabled: false,
      createdAt: null,
      updatedAt: null,
      fromNetworks: null,
      toNetworks: null,
      tokens: [],
    }
  }

  return {
    enabled: true,
    createdAt: bridge.created_at,
    updatedAt: bridge.updated_at,
    fromNetworks: bridge.from_networks,
    toNetworks: bridge.to_networks,
    tokens: bridge.tokens,
  }
}

export const GET = createApiEndpoint("getSiloBridge", async (_req, ctx) => {
  const bridge = await getSiloBridge(Number(ctx.params.id))

  if (!bridge) {
    return getBridgeSchema()
  }

  return getBridgeSchema(bridge)
})

export const POST = createApiEndpoint("createSiloBridge", async (_req, ctx) => {
  const silo = await getSilo(Number(ctx.params.id))

  if (!silo) {
    abort(404)
  }

  const existingBridge = await getSiloBridge(silo.id)

  if (existingBridge) {
    abort(400, "Bridge already exists")
  }

  const bridge = await createSiloBridge({ silo_id: silo.id })

  return getBridgeSchema(bridge)
})

export const PUT = createApiEndpoint("updateSiloBridge", async (_req, ctx) => {
  const siloId = Number(ctx.params.id)
  const bridge = await getSiloBridge(siloId)

  if (!bridge) {
    abort(404)
  }

  const { fromNetworks, toNetworks } = ctx.body

  const updatedBridge = await updateSiloBridge(siloId, {
    fromNetworks: fromNetworks ? getValidNetworks(fromNetworks) : undefined,
    toNetworks: toNetworks ? getValidNetworks(toNetworks) : undefined,
    tokens: ctx.body.tokens,
  })

  return getBridgeSchema(updatedBridge)
})
