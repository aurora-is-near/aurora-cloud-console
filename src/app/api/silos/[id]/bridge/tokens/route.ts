import { createApiEndpoint } from "@/utils/api"
import { abort } from "../../../../../../utils/abort"
import { getSilo } from "@/actions/silos/get-silo"
import { getSiloToken } from "@/actions/silo-tokens/get-silo-token"
import { updateToken } from "@/actions/tokens/update-token"
import { ApiResponseBody } from "@/types/api"
import { getSiloTokenByAddress } from "@/actions/silo-tokens/get-silo-token-by-address"
import { createToken } from "@/actions/tokens/create-token"

const bridgeToken = async (tokenId: number) => {
  const updatedToken = await updateToken(tokenId, {
    bridge_deployment_status: "PENDING",
  })

  return { status: updatedToken.bridge_deployment_status }
}

const bridgeExistingToken = async (
  siloId: number,
  tokenId: number,
): Promise<ApiResponseBody<"bridgeSiloToken">> => {
  const token = await getSiloToken(siloId, tokenId)

  if (!token) {
    abort(404)
  }

  if (token.bridge_deployment_status === "DEPLOYED") {
    abort(400, "Token is already deployed")
  }

  return bridgeToken(token.id)
}

const bridgeCustomToken = async (
  siloId: number,
  symbol: string,
  address: string,
): Promise<ApiResponseBody<"bridgeSiloToken">> => {
  const token = await getSiloTokenByAddress(siloId, address)

  if (token?.bridge_deployment_status === "DEPLOYED") {
    abort(400, "Token is already deployed")
  }

  if (token) {
    return bridgeToken(token.id)
  }

  const newToken = await createToken({
    symbol,
    address,
    deployment_status: "PENDING",
    bridge_deployment_status: "PENDING",
    silo_id: siloId,
    type: null,
    name: null,
    decimals: null,
    bridge_addresses: [],
    fast_bridge: false,
    bridge_origin: null,
    icon_url: null,
  })

  return { status: newToken.bridge_deployment_status }
}

export const POST = createApiEndpoint("bridgeSiloToken", async (_req, ctx) => {
  const silo = await getSilo(Number(ctx.params.id))
  const badRequestMessage = "Must provide either tokenId or symbol/address"

  if (!silo) {
    abort(404)
  }

  if (ctx.body.tokenId && (ctx.body.symbol || ctx.body.address)) {
    abort(400, badRequestMessage)
  }

  if (ctx.body.tokenId) {
    return bridgeExistingToken(silo.id, ctx.body.tokenId)
  }

  if (!ctx.body.symbol || !ctx.body.address) {
    abort(400, badRequestMessage)
  }

  return bridgeCustomToken(silo.id, ctx.body.symbol, ctx.body.address)
})
