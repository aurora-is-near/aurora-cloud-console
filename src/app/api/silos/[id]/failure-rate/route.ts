import { getSilo } from "@/actions/silos/get-silo"
import { abort } from "@/utils/abort"
import { createApiEndpoint } from "@/utils/api"
import { getFailureRateChart } from "@/utils/grafana/charts"

export const GET = createApiEndpoint(
  "getSiloFailureRate",
  async (_req, ctx) => {
    const silo = await getSilo(Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    const item = await getFailureRateChart({
      network: silo.grafana_network_key,
    })

    return { items: [item] }
  },
)
