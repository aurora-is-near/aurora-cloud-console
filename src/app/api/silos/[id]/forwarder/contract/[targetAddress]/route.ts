import { getAddress } from "ethers"
import { createApiEndpoint } from "@/utils/api"
import { abort } from "@/utils/abort"
import { forwarderApiClient } from "@/utils/forwarder-api/client"
import { getTeamSilo } from "@/actions/team-silos/get-team-silo"

export const GET = createApiEndpoint(
  "getForwarderAddress",
  async (_req, ctx) => {
    const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    let targetAddress

    try {
      targetAddress = getAddress(ctx.params.targetAddress)
    } catch {
      abort(400, "Invalid address")
    }

    const result = await forwarderApiClient.getContract({
      targetAddress,
      targetNetwork: silo.engine_account,
    })

    if (!result.result.address) {
      abort(404)
    }

    return {
      address: result.result.address,
    }
  },
)
