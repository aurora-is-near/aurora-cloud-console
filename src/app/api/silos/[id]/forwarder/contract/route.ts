import { getAddress } from "ethers"
import { createApiEndpoint } from "@/utils/api"
import { abort } from "@/utils/abort"
import { forwarderApiClient } from "@/utils/forwarder-api/client"
import { getTeamSilo } from "@/actions/team-silos/get-team-silo"

export const POST = createApiEndpoint(
  "createForwarderAddress",
  async (_req, ctx) => {
    const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    let targetAddress

    try {
      targetAddress = getAddress(ctx.body.targetAddress)
    } catch {
      abort(400, "Invalid address")
    }

    const result = await forwarderApiClient.createContract({
      targetAddress,
      targetNetwork: silo.engine_account,
    })

    return {
      address: result.result.address,
    }
  },
)
