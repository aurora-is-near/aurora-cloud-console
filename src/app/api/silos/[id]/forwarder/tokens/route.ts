import { JsonRpcProvider } from "ethers"
import { createApiEndpoint } from "@/utils/api"
import { abort } from "@/utils/abort"
import { getTeamSilo } from "@/actions/team-silos/get-team-silo"
import { forwarderApiClient } from "@/utils/forwarder-api/client"
import { Silo } from "@/types/types"
import { ForwarderTokenSymbol } from "@/types/forwarder-tokens"
import { checkTokenBySymbol } from "@/utils/check-token-contract"

type Token = {
  address: string
  decimals: number
  symbol: ForwarderTokenSymbol
}

// These are the tokens potentially supported by the forwarder. The address is
// the token's NEAR contract address.
const KNOWN_TOKENS: Token[] = [
  {
    address: "near",
    decimals: 24,
    symbol: "NEAR",
  },
  {
    address: "wrap.near",
    decimals: 24,
    symbol: "wNEAR",
  },
  {
    address: "usdt.tether-token.near",
    decimals: 6,
    symbol: "USDt",
  },
  {
    address: "17208628f84f5d6ad33f0da3bbbeb27ffcb398eac501a31bd6ad2011e36133a1",
    decimals: 6,
    symbol: "USDC",
  },
  {
    address: "aaaaaa20d9e0e2461697782ef11675f668207961.factory.bridge.near",
    decimals: 18,
    symbol: "AURORA",
  },
]

const isKnownToken = (symbol: string): symbol is ForwarderTokenSymbol => {
  return KNOWN_TOKENS.some((known) => known.symbol === symbol)
}

const checkTokenContracts = async (silo: Silo, tokens: string[]) => {
  const provider = new JsonRpcProvider(silo.rpc_url)

  await Promise.all(
    tokens.map(async (symbol) => {
      if (!isKnownToken(symbol)) {
        abort(400, `Unknown token: ${symbol}`)
      }

      if (!(await checkTokenBySymbol(provider, symbol))) {
        abort(400, `Token contract not deployed: ${symbol}`)
      }
    }),
  )
}

const addTokens = async (silo: Silo, tokens: string[]) => {
  if (!tokens.length) {
    return
  }

  await forwarderApiClient.addSupportedToken({
    target_network: silo.engine_account,
    tokens: tokens.map((symbol) => {
      const { address, decimals } =
        KNOWN_TOKENS.find((token) => token.symbol === symbol) ?? {}

      if (!address || !decimals) {
        abort(400, `Unknown token: ${symbol}`)
      }

      return { address, decimals, symbol }
    }),
  })
}

const removeTokens = async (silo: Silo, tokens: string[]) => {
  if (!tokens.length) {
    return
  }

  await forwarderApiClient.removeSupportedToken({
    target_network: silo.engine_account,
    token_addresses: tokens.map((symbol) => {
      const { address } =
        KNOWN_TOKENS.find((token) => token.symbol === symbol) ?? {}

      if (!address) {
        abort(400, `Unknown token: ${symbol}`)
      }

      return address
    }),
  })
}

export const GET = createApiEndpoint(
  "getForwarderTokens",
  async (_req, ctx) => {
    const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    const provider = new JsonRpcProvider(silo.rpc_url)
    const {
      result: { tokens: supportedTokens },
    } = await forwarderApiClient.getSupportedTokens({
      target_network: silo.engine_account,
    })

    return {
      items: await Promise.all(
        KNOWN_TOKENS.map(async ({ symbol, decimals }) => ({
          symbol,
          decimals,
          contractDeployed: await checkTokenBySymbol(provider, symbol),
          enabled: (supportedTokens ?? []).some(
            (token) => token.symbol === symbol,
          ),
        })),
      ),
    }
  },
)

export const POST = createApiEndpoint(
  "addForwarderTokens",
  async (_req, ctx) => {
    const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    const { tokens } = ctx.body

    await checkTokenContracts(silo, tokens)
    await addTokens(silo, tokens)

    return {
      status: "ok",
    }
  },
)

export const DELETE = createApiEndpoint(
  "removeForwarderTokens",
  async (_req, ctx) => {
    const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    const { tokens } = ctx.body

    await checkTokenContracts(silo, tokens)
    await removeTokens(silo, tokens)

    return {
      status: "ok",
    }
  },
)

export const PUT = createApiEndpoint(
  "updateForwarderTokens",
  async (_req, ctx) => {
    const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

    if (!silo) {
      abort(404)
    }

    const { tokens } = ctx.body

    const {
      result: { tokens: currentTokens },
    } = await forwarderApiClient.getSupportedTokens({
      target_network: silo.engine_account,
    })

    const currentTokenSymbols = (currentTokens ?? []).map(
      (token) => token.symbol,
    )

    const tokensToRemove = currentTokenSymbols.filter(
      (symbol) => !tokens.includes(symbol),
    )

    const tokensToAdd = tokens.filter(
      (symbol) => !currentTokenSymbols.includes(symbol),
    )

    await checkTokenContracts(silo, tokensToAdd)

    await Promise.all([
      addTokens(silo, tokensToAdd),
      removeTokens(silo, tokensToRemove),
    ])

    return {
      status: "ok",
    }
  },
)
