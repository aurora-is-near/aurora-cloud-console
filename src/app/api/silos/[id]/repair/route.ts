import { createApiEndpoint } from "@/utils/api"
import { abort } from "@/utils/abort"
import { repairSilo } from "@/utils/repair"
import { getTeamSilo } from "@/actions/team-silos/get-team-silo"

export const POST = createApiEndpoint("repair", async (_req, ctx) => {
  const silo = await getTeamSilo(ctx.team.id, Number(ctx.params.id))

  if (!silo) {
    abort(404)
  }

  return repairSilo(silo)
})
