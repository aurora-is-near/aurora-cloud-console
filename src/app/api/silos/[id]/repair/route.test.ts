/**
 * @jest-environment node
 */
import { ethers } from "ethers"
import { POST } from "./route"
import {
  createInsertOrUpdate,
  createSelect,
  mockSupabaseClient,
} from "../../../../../../test-utils/mock-supabase-client"
import { createMockSilo } from "../../../../../../test-utils/factories/silo-factory"
import { invokeApiHandler } from "../../../../../../test-utils/invoke-api-handler"
import { setupJestOpenApi } from "../../../../../../test-utils/setup-jest-openapi"

const currentTime = new Date("2025-03-17T09:00:00Z")

jest.useFakeTimers()
jest.setSystemTime(currentTime)

jest.mock("ethers")
jest.mock("@/utils/contract-changer-api/contract-changer-api-client", () => ({
  contractChangerApiClient: {
    mirrorErc20Token: jest.fn(({ token }) => ({
      tx_hash: `mock_tx_hash_${token}`,
    })),
    setBaseToken: jest.fn(() => ({ tx_hash: "mock_set_base_token_tx_hash" })),
    makeStorageDeposit: jest.fn(() => ({
      tx_hash: "mock_storage_deposit_tx_hash",
    })),
  },
}))

jest.mock("near-api-js", () => ({
  ...jest.requireActual("near-api-js"),
  Account: jest.fn(() => ({
    getAccountBalance: jest.fn(() => ({
      total: "33137693971864085399999999",
      available: "18370803971864085399999999",
      staked: "0",
    })),
    viewFunction: jest.fn(() => ({
      total: "12345678901234567890",
      available: "42",
    })),
  })),
  connect: jest.fn(() => ({
    connection: {
      provider: {
        txStatus: jest.fn(() => ({ status: { SuccessValue: "" } })),
      },
    },
  })),
}))

jest.mock("../../../../../utils/api", () => ({
  createApiEndpoint: jest.fn((_name, handler) => handler),
}))

describe("Silo repair route", () => {
  beforeAll(setupJestOpenApi)

  beforeEach(() => {
    mockSupabaseClient.from("silos").select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("silos")
      .update.mockReturnValue(createInsertOrUpdate({}))
    mockSupabaseClient
      .from("silo_config_transactions")
      .select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("silo_config_transactions")
      .update.mockReturnValue(createInsertOrUpdate([]))
    mockSupabaseClient
      .from("bridged_token_requests")
      .select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("bridged_tokens")
      .select.mockReturnValue(createSelect([]))
    mockSupabaseClient
      .from("silo_bridged_tokens")
      .select.mockReturnValue(createSelect([]))
    ;(ethers.Contract as jest.Mock).mockImplementation(() => ({
      symbol: () => {
        throw new Error("Not implemented")
      },
    }))
  })

  it("repairs a silo", async () => {
    const oneHourAgo = new Date(currentTime.getTime() - 60 * 60 * 1000)

    const mockSilo = createMockSilo({
      is_active: false,
      inspected_at: oneHourAgo.toISOString(),
      base_token_symbol: "AURORA",
    })

    mockSupabaseClient
      .from("silos")
      .select.mockReturnValue(createSelect(mockSilo))

    const res = await invokeApiHandler("POST", "/api/silos/1/repair", POST)

    expect(res.status).toBe(200)
    expect(res).toSatisfyApiSpec()
    expect(res.body).toMatchSnapshot()
  })
})
