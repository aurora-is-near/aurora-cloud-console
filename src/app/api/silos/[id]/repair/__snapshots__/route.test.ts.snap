// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Silo repair route repairs a silo 1`] = `
{
  "initialisation": {
    "baseTokenSet": "PENDING",
    "defaultTokensDeployed": "PENDING",
    "isSiloActive": false,
  },
  "pendingBridgedTokens": {
    "numberOfPendingDeployments": 0,
    "numberOfPendingDeploymentsResolved": 0,
    "numberOfRequests": 0,
    "numberOfRequestsResolved": 0,
  },
  "pendingTransactions": {
    "numberOfPendingTransactions": 0,
    "numberOfPendingTransactionsResolved": 0,
  },
  "status": "ok",
}
`;
